src/NumericalAlgorithms/LinearSolver/LinearSolver.hpp

```diff
@@ -0,0 +1,279 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <memory>
+#include <optional>
+#include <pup.h>
+#include <pup_stl.h>
+#include <type_traits>
+#include <utility>
+
+#include "NumericalAlgorithms/Convergence/HasConverged.hpp"
+#include "Options/Auto.hpp"
+#include "Options/Options.hpp"
+#include "Parallel/CharmPupable.hpp"
+#include "Parallel/PupStlCpp17.hpp"
+#include "Utilities/FakeVirtual.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/Registration.hpp"
+#include "Utilities/Requires.hpp"
+
+namespace LinearSolver::Serial {
+
+/// Indicates the linear solver uses no preconditioner. It may perform
+/// compile-time optimization for this case.
+struct NoPreconditioner {};
+
+/// Indicates the linear solver uses any other registered linear solver as
+/// preconditioner.
+struct AnyPreconditioner {};
+
+/// Registrars for linear solvers
+namespace Registrars {}
+
+/*!
+ * \brief Base class for serial linear solvers that supports factory-creation
+ * and nested preconditioning.
+ *
+ * To enable support for preconditioning in your derived linear solver class,
+ * pass any type that has a `solve` and a `reset` function as the
+ * `Preconditioner` template parameter. It can also be an abstract
+ * `LinearSolver` type, which means that any other linear solver can be used as
+ * preconditioner. Pass `NoPreconditioner` to disable support for
+ * preconditioning.
```

Comment:
I'm confused.. can you clarify what the `Preconditioner` template parameter is here? I don't see it in the `LinearSolver` class below... can you clarify what you mean in this comment? Or perhaps, is this comment supposed to attach to the `PreconditionedLinearSolver` class further down?

Reply:
it was indeed intended for `PreconditionedLinearSolver`, thanks for catching this 👍